source/server/listener_manager_impl.cc

```diff
@@ -694,20 +705,49 @@ void ListenerManagerImpl::onListenerWarmed(ListenerImpl& listener) {
   updateWarmingActiveGauges();
 }
 
-void ListenerManagerImpl::drainFilterChains(ListenerImplPtr&& listener) {
+void ListenerManagerImpl::onIntelligentListenerWarmed(ListenerImpl& listener) {
```

Comment:
Lots of copying in this function. Having been through the change in more detail, my general impression is that it will be simpler to perform an in place update of an existing listener. This would preserve the tag and I think the draining of filter chains could be done within the listener itself?

Reply:
> I think the draining of filter chains could be done within the listener itself?

That is the consequence of reusing tag, correct?

Reply:
yeah I think if we have everything happening within a ListenerImpl some things get a lot easier (sharing tag, some state management) but some things get more difficult. I would definitely be curious to see what it looks like to update the ListenerImpl in place as my intuition is that the overall code will wind up simpler to understand and reason about.